libc/pthread.c

```diff
@@ -41,39 +41,28 @@
 #undef pthread
 #endif
 
-OE_STATIC_ASSERT(sizeof(struct __pthread) <= sizeof(((td_t*)NULL)->pthread));
+OE_STATIC_ASSERT(sizeof(struct __pthread) <= OE_THREAD_LOCAL_SPACE);
 OE_STATIC_ASSERT(sizeof(pthread_once_t) == sizeof(oe_once_t));
 OE_STATIC_ASSERT(sizeof(pthread_spinlock_t) == sizeof(oe_spinlock_t));
 OE_STATIC_ASSERT(sizeof(pthread_mutex_t) >= sizeof(oe_mutex_t));
 OE_STATIC_ASSERT(sizeof(pthread_cond_t) >= sizeof(oe_cond_t));
 OE_STATIC_ASSERT(sizeof(pthread_rwlock_t) >= sizeof(oe_rwlock_t));
 
+static __thread struct __pthread _pthread_self;
+
 static void _pthread_self_init()
 {
-    td_t* td = oe_get_td();
-
-    if (td)
-    {
-        struct __pthread* self = (struct __pthread*)td->pthread;
-        memset(self, 0, sizeof(struct __pthread));
-        self->locale = C_LOCALE;
-    }
+    memset(&_pthread_self, 0, sizeof(struct __pthread));
+    _pthread_self.locale = C_LOCALE;
```

Comment:
If all that _pthread_self_init is doing is to zero out memory and set one field, then we could just do:
static __thread struct __pthread _pthread_self = { .locale = C_LOCALE };

We already use the C99  initializer syntax (e.g oe_sgx_properties).

The only difference between initializing it via a delayed function call and explicitly initializing via C99 initializer (or otherwise) is that in the former, the tls variable will be put in the .tbss section; in the latter it will be put in the .tdata section. It doesn't really matter for us where the variable is put.

Reply:
Oh, good point! It slipped my mind that Mike already used the designator syntax before and I was thinking we would need to avoid new instances of C99 use given Dave's request. This is a good instance for its use though and it's much less public than the `oe_sgx_properties_t` initializer.